### PR TITLE
[model,ops] feat: add DeepSeek V4 TileLang kernels

### PR DESCRIPTION
**Pull request by Luosuu**
### What does this PR do?

Follow-up to #902. Port the DeepSeek V4 model-specific kernel library from [radixark/miles](https://github.com/radixark/miles/tree/main/miles_plugins/models/deepseek_v4/ops/kernel) and wire the TileLang kernels into VeOmni's Transformers DeepSeek V4 model path.

This branch also includes the four checkpoint-loading fixes from #902 so the official `deepseek-ai/DeepSeek-V4-Flash` checkpoint can be loaded for end-to-end validation.

### Checklist Before Starting

- Related work: #902
- Upstream kernel reference: radixark/miles at `06c9a5c7b9a28507eb05f1173be0b038c276a361`
- PR title follows `[{modules}] {type}: {description}` format.

### Design & Code Changes

- Add model-specific ops under `veomni/ops/kernels/deepseek_v4/`:
  - TileLang sparse-attention forward kernel.
  - TileLang Lightning Indexer forward/backward kernels.
  - FP8 activation quantization.
  - Precision-aligned BF16 linear autograd.
- Add DeepSeek V4 model dispatch through `OpsImplementationConfig`:
  - `dsa_indexer_backend: eager | cudnn | tilelang`
  - `dsa_attention_backend: eager | flashmla_cudnn | tilelang_sparse`
- Dispatch supported CUDA BF16 prefill shapes to TileLang and retain eager fallback for unsupported dtype, shape, position, dropout, or layout combinations.
- Regenerate the Transformers v5 DeepSeek V4 patch through patchgen; generated files were not edited manually.
- Preserve Miles Apache-2.0 attribution and DeepSeek's MIT notice for the activation quantizer.
- Harden the imported kernels against padded/out-of-range indices, partial tiles, padded heads, unsupported indexer head counts, and non-contiguous autograd gradients.
- Include #902 checkpoint conversion fixes for official DeepSeek V4 Flash names, packed FP4 weights, prefixed keys, MTP exclusions, and converter prefix targeting.

TileLang remains lazily imported so CPU and NPU environments can import VeOmni without loading the optional GPU dependency. TileLang covers the DSA indexer and sparse-attention kernels; EP dispatch and expert computation use VeOmni's fused Triton MoE backend.

### Validation

Development environment is managed from the repository `pyproject.toml` with uv.

- `uv sync --locked --extra gpu --dev`
- `make quality`
- `patchgen --check`: all 26 patch configurations passed
- `.venv/bin/pytest -q tests/ops/test_deepseek_v4_kernels.py`: 9 passed on NVIDIA GB200 (SM100)
- `.venv/bin/pytest -q tests/ops/test_moe_hw_gate.py`: 17 passed
- `.venv/bin/pytest -q tests/models/test_models_patch.py -k deepseek_v4`: 1 passed
- `.venv/bin/pytest -q tests/models/test_checkpoint_tensor_converter.py -k DeepseekV4`: 73 passed, 51 deselected
- Toy eager-vs-TileLang model forward: same top-1 token, cosine similarity `0.9999161959`

The kernel tests cover non-power-of-two top-k, non-block-aligned sequence/KV sizes, bounded invalid indices, padded heads, non-contiguous backward gradients, and FP8 normal/in-place/E8M0 scaling. NVIDIA `compute-sanitizer` was not available on the host.

### Real-weight prefill logits comparison

Compared VeOmni against the official [`deepseek-ai/DeepSeek-V4-Flash/inference`](https://huggingface.co/deepseek-ai/DeepSeek-V4-Flash/tree/main/inference) implementation using the full published checkpoint.

- Weights: complete 46-shard snapshot, approximately 149 GiB; the official inference path used its MP=4 converted checkpoints.
- Hardware: 4 x NVIDIA GB200.
- Prompt: `Hello, who are you?` encoded with official `encode_messages(..., thinking_mode="chat")`, producing 10 identical input tokens in both paths.
- Reference path: official `inference/model.py` `Transformer`, BF16, MP=4.
- VeOmni path: Transformers v5 patch, BF16, FSDP2, EP=4 (64/256 experts per rank), `moe_implementation=fused_triton`, `dsa_indexer_backend=tilelang`, and `dsa_attention_backend=tilelang_sparse`.
- Comparison target: final prefill-token logits over the full 129,280-token vocabulary.

| Metric | Result |
| --- | ---: |
| Input IDs identical | yes |
| All logits finite | yes |
| Cosine similarity | **0.9855524898** |
| Top-1 token | **19923 in both paths** |
| Top-10 overlap | **8/10** |
| Mean absolute error | 0.7263982 |
| Maximum absolute error | 4.0532575 |
| RMSE | 0.9059990 |

The real-weight VeOmni forward completed successfully with TileLang DSA kernels enabled.

### API and Usage Example

```python
from veomni.ops.kernels.deepseek_v4 import (
    act_quant,
    linear_bf16_fp32,
    sparse_attn_tilelang,
    v4_lighting_indexer,
)
```

### Checklist Before Submitting

- [x] Read the Contribute Guide
- [x] Applied pre-commit checks
- [x] Added/updated documentation
- [x] Added focused GPU tests
- [x] Loaded the full official checkpoint and compared real-weight prefill logits

### CODE REVIEW

**Comment by Luosuu** [comment]
Real-weight prefill logits comparison (DeepSeek-V4-Flash)

- Base: rebased onto `github/main` at `7b0cff8`; branch result at `00caefe`
- Weights: full `deepseek-ai/DeepSeek-V4-Flash` snapshot (46 safetensor shards, ~149 GiB); official inference conversion used MP=4 checkpoints
- Hardware: 4x NVIDIA GB200
- Prompt: `Hello, who are you?` via the official `encode_messages(..., thinking_mode="chat")` path (10 input tokens)
- Reference: upstream `inference/` implementation
- VeOmni: FSDP2, EP=4 (64/256 experts per rank), `moe_implementation=fused_triton`, `dsa_indexer_backend=tilelang`, `dsa_attention_backend=tilelang_sparse`, BF16
- Compared logits: final prefill token, full vocabulary `[129280]`

Post-rebase results:
- Input IDs identical: yes
- All logits finite: yes
- Cosine similarity: **0.9847503901**
- Top-1 token: **19923** for both
- Top-10 overlap: **8/10**
- Mean absolute error: **0.7380947**
- Max absolute error: **4.8393769**
- RMSE: **0.9314570**

Before rebase:
- Cosine similarity: **0.9855524898**
- MAE: **0.7263982**
- Max error: **4.0532575**
- RMSE: **0.9059990**

Conclusion: the gap did **not** become smaller after rebasing. Cosine decreased by `0.00080210`; top-1 and top-10 overlap were unchanged. The relevant upstream change passes `swiglu_limit=10.0` into VeOmni fused MoE, matching both the Hugging Face and official inference configs, so the rebased path is semantically aligned even though this aggregate similarity metric did not improve. TileLang covers the DSA indexer and sparse-attention kernels; EP dispatch and expert computation use VeOmni fused Triton MoE.

File: veomni/ops/kernels/deepseek_v4/tilelang_indexer_bwd.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+                for i in T.Parallel(block_I):
+                    if bi_i * block_I + i < topk:
+                        indices_shared[i] = TopkIndices[bx, bi_i * block_I + i]
+                        grad[i] = OGrad[bx, bi_i * block_I + i]
+
+                T.sync_threads()
+                for i, j in T.Parallel(block_I, dim):
+                    index_k_shared[i, j] = T.if_then_else(
+                        indices_shared[i] > -1 and indices_shared[i] < seq_len, IndexK[indices_shared[i], j], 0
+                    )
```
[the remark this answers is not in the document]
Comment (Luosuu) [reply veomni/ops/kernels/deepseek_v4/tilelang_indexer_bwd.py:103]: Verified against the TileLang version used by this branch: this claim is not applicable. `T.if_then_else` lowers to `tir.if_then_else`, whose API explicitly states that the unselected branch is not executed and that it can guard out-of-bounds access. TileLang also recognizes the `if_then_else(predicate, global_load, 0)` pattern as a predicated load during lowering. The GB200 forward/backward test exercises `-1`, `-2`, and upper-out-of-range indices successfully. Therefore the existing load is guarded and an additional `safe_idx` is not required.

File: veomni/ops/kernels/deepseek_v4/tilelang_indexer.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        cu_seqlen_ks, cu_seqlen_ke = _make_causal_cu_seqlens(seqlen_q, seq_len_kv, compress_ratio, index_q.device)
+
+        # [batch, seqlen, seqlen_kv]
+        logits = batched_indexer_fwd(index_q, index_k, weights, cu_seqlen_ks, cu_seqlen_ke)
+
+        if topk_indices is None:
+            actual_topk = min(topk, seq_len_kv)
+            index_score, topk_indices = torch.topk(logits, actual_topk, dim=-1)
+            topk_indices = topk_indices.to(torch.int32)
+            topk_indices = topk_indices.masked_fill(index_score == -torch.inf, -1)
```
[the remark this answers is not in the document]
Comment (Luosuu) [reply veomni/ops/kernels/deepseek_v4/tilelang_indexer.py:82]: This is a valid standalone API-shape observation, but it cannot cause the described downstream model crash in the current integration. `DeepseekV4Indexer.forward` computes `top_k = min(self.index_topk, compressed_len)` before calling `v4_lighting_indexer`, so the model never requests `topk > seq_len_kv`. For a direct caller that does request it, the function currently returns the actual available width, while the docstring says `[... , topk]`; padding with `-1` would make that public contract stricter. We acknowledge this as a low-priority API consistency edge case, separate from model-path correctness.

File: veomni/ops/kernels/deepseek_v4/tilelang_indexer_fwd.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    block_n = 256
+    padded_seq_len = (seq_len + block_q - 1) // block_q * block_q
+    padded_seq_len_kv = (seq_len_kv + block_n - 1) // block_n * block_n
+    if padded_seq_len != seq_len:
+        q = torch.cat([q, q.new_zeros(padded_seq_len - seq_len, heads, index_dim)], dim=0)
+        weights = torch.cat([weights, weights.new_zeros(padded_seq_len - seq_len, heads)], dim=0)
+        cu_seqlen_ks = torch.cat([cu_seqlen_ks, cu_seqlen_ks.new_full((padded_seq_len - seq_len,), seq_len_kv)])
+        cu_seqlen_ke = torch.cat([cu_seqlen_ke, cu_seqlen_ke.new_full((padded_seq_len - seq_len,), seq_len_kv)])
+    if padded_seq_len_kv != seq_len_kv:
+        kv = torch.cat([kv, kv.new_zeros(padded_seq_len_kv - seq_len_kv, index_dim)], dim=0)
```
[the remark this answers is not in the document]
Comment (Luosuu) [reply veomni/ops/kernels/deepseek_v4/tilelang_indexer_fwd.py:177]: The concern is reasonable as defensive hardening, but it is not a correctness issue in the supported VeOmni path. `indexer_fwd_interface` is an internal low-level helper and its only in-tree caller, `batched_indexer_fwd`, passes `k[:, b, :].contiguous()` (and likewise normalizes `q` and `weights`). The DeepSeek-V4 model dispatch also passes contiguous transposed inputs before reaching the wrapper. Thus a non-contiguous `kv` cannot reach the TileLang kernel through the current model path. We are keeping the caller-side layout contract rather than adding a redundant copy here.

File: veomni/ops/kernels/deepseek_v4/tilelang_sparse_mla_fwd.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        kv:        [B, S_kv, D] bf16
+        attn_sink: [H] fp32
+        topk_idxs: [B, S, topk] int32
+        sm_scale:  float or None (defaults to 1/sqrt(D))
+
+    Returns:
+        out: [B, S, H, D] bf16
+        lse: [B, S, H] fp32
+    """
+    assert q.is_contiguous() and kv.is_contiguous() and topk_idxs.is_contiguous()
```
[the remark this answers is not in the document]
Comment (Luosuu) [reply veomni/ops/kernels/deepseek_v4/tilelang_sparse_mla_fwd.py:190]: The production model path already calls `sparse_attn_tilelang` with contiguous `query`, `key`, sinks, and indices, so sequence/tensor-parallel views do not reach this assertion. The assertion produces a deterministic contract failure rather than undefined kernel behavior. Also, the suggested change in this low-level forward helper would be incomplete for autograd: `DeepSeekV4SparseAttention.forward` would still save the original non-contiguous tensors, and the backward interface enforces the same contiguous contract. Supporting arbitrary views would require normalization in the outer autograd Function before both execution and `save_for_backward`, plus a forward/backward regression test. This is therefore optional API hardening, not a current production crash.

**Comment by Luosuu** [comment]
Packed dynamic-batch update (commit 70b2bd3):

- The TileLang indexer now accepts per-query compressed-KV start/end ranges, so packed input stays on the TileLang path without an eager fallback.
- HCA/CSA/indexer compression windows, RoPE positions, indexer ranges, and ordinary sliding/causal attention are isolated at packed-sample boundaries.
- Packing metadata (window indices, causal ranges, and the reusable HCA bias) is built once per model forward. Per layer, only the layer-specific projected K/G gather remains for autograd correctness.
- Packed training disables decode-cache accumulation so the path is correct both with and without gradient checkpointing.

Local GB200 validation:

- 4-GPU FSDP2 packed dyn-bsz smoke, 2x1024, gradient checkpointing enabled: 2 full training steps including backward passed.
- 4-GPU FSDP2 packed dyn-bsz smoke, 4x512, gradient checkpointing disabled: 2 full training steps including backward passed.
- Combined packed smoke result: 2 passed in 38.60s; final no-GC loss 6.98719931 -> 6.94665861 and grad norm 0.80410558 -> 0.79678643.
- DeepSeek V4 GPU ops/model tests: 14 passed, including packed indexer forward/backward versus the PyTorch reference.
- make quality, make check-patchgen (26 configs), and git diff --check passed.

The tests are skipped on L20/SM89 and ROCm because these TileLang kernels require NVIDIA SM90+. A fused packed-compression kernel could reduce the remaining gather temporaries, but the current metadata caching removes the repeated per-layer index/range construction.

**Comment by Luosuu** [comment]
## TileKernels mHC backend update

Commit `454f8b4` adds a uv-managed `tile-kernels==1.0.0` dependency and wires DeepSeek V4 mHC through three registry-backed `OpSlot`s selected by `model.ops_implementation.mhc_backend=tile_kernels`:

- mHC pre projection, Sinkhorn normalization, and stream collapse
- residual post-mix after attention and MLP
- final hyper-head stream collapse

The backend is explicit (no eager fallback after selection), requires BF16 activations, `hc_mult=4`, and NVIDIA SM90+. Training uses TileKernels forward/backward kernels with independent autograd accumulation; the fused no-grad pre path is also covered.

TileKernels requires TileLang 0.1.9. That upgrade exposed a stricter software-pipeline dependency check in the existing bounds-safe sparse-MLA indirect gather. The gather loop is now explicitly unpipelined (`num_stages=0`), preserving safe invalid-index handling; its forward/backward numerical test passes on 0.1.9.

Validation on 4x NVIDIA GB200 (SM100):

- `tests/ops/test_deepseek_v4_kernels.py tests/ops/test_mhc_tile_kernels.py`: **20 passed**
- mHC pre/post/head forward and backward vs eager: `rtol=2e-2`, `atol=2e-2`, gradient cosine similarity `>0.98`; Sinkhorn repeat count 20
- fused no-grad mHC pre path vs eager: passed
- packed dynamic-batch full-model training/backward, 4 GPUs:
  - packed 2x1024 with gradient checkpointing: passed
  - packed 4x512 without gradient checkpointing: passed
  - combined result: **2 passed in 73.58s**
- patchgen check: all 26 generated configs up to date
- `make quality`: passed
- `uv lock --check`: resolved 234 packages successfully
- FlashQLA import API compatibility with TileLang 0.1.9 was checked by bypassing its SM90-only architecture gate; runtime execution could not be tested on SM100 because FlashQLA itself rejects non-SM90 devices.

The DeepSeek V4 TileLang/TileKernels E2E cases remain skipped on L20 CI because all optimized paths require SM90 or newer.

**Review by FoolPlayer — APPROVED** [review]
